_posts/2021-09-03-rudiments-of-data-wrangling-in-rust.md

```diff
@@ -0,0 +1,514 @@
+---
+
+The [Rust](https://www.rust-lang.org) programming 
+language has many advantages over 
+its contemporary popular alternatives: memory safety, 
+static typing with powerful type inference, algebraic data 
+types, fearless concurrency, a friendly community, and 
+the list goes on.
+
+Rust is typically positioned as a language for "systems programming",
+which usually means something rather close to the operating system,
+such as networking or device drivers.
+However, Rust is a general-purpose programming language, and there is
+nothing stopping you from using it for tasks that often are done in
+Python.
+
+Since I started really diving into Rust in the spring of 2021, I've
+gradually become more fluent with it, to the extent that I can now
+create some utilities in Rust in many situations where I would normally 
+reach for Python.
+
+In this blog post I would like to demonstrate the potential of Rust
+by presenting my solution to a quick (and maybe a little dirty) 
+data wrangling task. It involves reading and parsing a large CSV file, 
+and then extracting rows that meet certain criteria. It's nothing 
+that couldn't be done in Python, or even with a shell script in Unix,
+but hopefully it serves as an example of how these kinds of common 
+tasks could be done in Rust, and as a basis for extension.
+
+The Rust program presented here can be found on 
+[GitHub](https://github.com/solita-jereka/vehicles-rs), so that
+you can clone it or fork it, run it on your own machine and experiment
+further.
+
+## Some basic Rust knowledge
+
+If you haven't already picked up the basics of Rust programming, I'd
+recommend to have a look at the 
+[Getting Started](https://www.rust-lang.org/learn/get-started) section of the Rust
+website. That will give you the information about how to download
+the necessary tools, and how to compile and run the Rust version 
+of the traditional first "Hello, world!" program.
+
+To learn a little more of the Rust language, the canonical book 
+["The Rust Programming Language"](https://doc.rust-lang.org/book/)
+(often know just as "the book")
+is almost mandatory reading. Rust has many characteristics that 
+are somewhat familiar to anyone who has programmed in Java, C#, 
+C++, or Python, but it also has many unique features borrowed 
+from slightly less mainstream programming languages, for a good 
+reason.
+
+![Rudiments of the Rust programming language (actually named after a fungus)](/img/rudiments-of-data-wrangling-in-rust/rusty-chain.png)
+
+*Photo by <a href="https://unsplash.com/@zmachacek?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText" target="_blank">Zdeněk Macháček</a> on <a href="https://unsplash.com/s/photos/rust?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText" target="_blank">Unsplash</a>*
+
+The most challenging part of learning Rust is its memory management,
+which is based on reference counting and strict control over memory
```

Comment:
Isn't reference counting more like an escape hatch for when you can't make the static ownership tracking system work for you? So the RC mention could be moved later in the paragraph, to come after theborrow checking etc explanation.

Reply:
Thanks, I thought I had fixed that already -- based on earlier feedback, decided to change "based on reference counting" to "based on tracking ownership". RC is more of an implementation detail.